crates/turbopack-core/src/resolve/mod.rs

```diff
@@ -2186,7 +2219,25 @@ async fn resolve_module_request(
         return Ok(result);
     }
 
-    let mut results = vec![];
+    // Self references, if the nearest package.json has the name of the requested
+    // module. This should match only using the exports field and no other
+    // fields/fallbacks.
+    if let FindSelfReferencePackageResult::Found { name, package_path } =
+        &*find_self_reference(lookup_path).await?
+    {
+        if name == module {
+            let result = resolve_into_package(
+                Value::new(path.clone()),
+                *package_path,
+                query,
+                fragment,
+                options,
+            );
+            if !(*result.is_unresolveable().await?) {
```

Comment:
Should we continue to look for other packages with the same name or should this better fail resolving here if the self reference doesn't resolve?

Reply:
I think that's an uncommon case, but https://nodejs.org/api/modules.html#all-together would continue searching. So I'm inclined to continue as well